feat(core): allow for injector to be specified when creating an embedded view

crisbeto opened the pull request:
Adds support for passing in an optional injector when creating an embedded view through `ViewContainerRef.createEmbeddedView` and `TemplateRef.createEmbeddedView`. The injector allows for the DI behavior to be customized within the specific template.

Fixes #14935.

jelbourn commented on packages/core/src/render3/chained_injector.ts:
```diff
@@ -0,0 +1,32 @@
+import {ProviderToken} from '../di/provider_token';
+import {NOT_FOUND_CHECK_ONLY_ELEMENT_INJECTOR} from '../view/provider_flags';
+
+export class ChainedInjector implements Injector {
```
Add a JsDoc class description? Probably also one for `get` that mentions the different from the base implementation?

AndrewKushnir commented on goldens/public-api/core/core.md:
```diff
@@ -1383,7 +1383,7 @@ export abstract class ViewContainerRef {
     // @deprecated
     abstract createComponent<C>(componentFactory: ComponentFactory<C>, index?: number, injector?: Injector, projectableNodes?: any[][], ngModuleRef?: NgModuleRef<any>): ComponentRef<C>;
-    abstract createEmbeddedView<C>(templateRef: TemplateRef<C>, context?: C, index?: number): EmbeddedViewRef<C>;
+    abstract createEmbeddedView<C>(templateRef: TemplateRef<C>, context?: C, index?: number, injector?: Injector): EmbeddedViewRef<C>;
```
+1, that's a great idea and this would make this API consistent with the [ViewContainerRef.createComponent](https://angular.io/api/core/ViewContainerRef#createComponent) API. We could also deprecate the current API with positional args at the same time.

crisbeto replied:
I was considering this, but the problem is that the `context` is usually an object literal that can contain anything and it'll be difficult to distinguish the two signatures at runtime. E.g. somebody could be doing `createEmbeddedView(template, {context: {}}`.

jelbourn replied:
Maybe this is wild, but what if we just introduced another function name? Something like `renderTemplate`? `createEmbeddedView` has always been a confusing name anyway, maybe this could be a good opportunity to do something more intuitive.

crisbeto replied:
That's also weird IMO, because it does exactly the same as `createEmbeddedView`.

jelbourn replied:
In that scenario we would deprecate `createEmbeddedView`. IMO the biggest (and extremely compelling) reason _not_ to do that would be all of the existing content (tutorials, videos, etc.) that reference `createEmbeddedView`. 

Curious to know what @alxhub might think here

AndrewKushnir replied:
Creating a new function is an interesting idea. If we were to do that, we should consider making the function standalone (not a part of the `ViewContainerRef` class), so that it gets tree-shaking benefits.

With standalone components, there will be a need to introduce a new bootstrap function, which would use a Component as a main entity (vs our current bootstrap APIs that work with NgModules). That new bootstrap function will contain a code to render a component and append to a specified DOM node, so this function (or a subset of the logic, extracted to a separate function) might also be used as a more ergonomic alternative to the `ViewContainerRef.createComponent` function (there will be no need to deal with the `ViewContainerRef` in most cases).

So the end state might look like this:

- renderComponent (standalone function, replacement for `ViewContainerRef.createComponent`)
- renderTemplate (standalone function, replacement for `ViewContainerRef.createEmbeddedView`)
- ViewContainerRef without "createComponent" and "createEmbeddedView" functions, i.e. a class that is exclusively focused on the view management (add/move/remove/retrieve a view)

Just some ideas for further discussions :)

AndrewKushnir commented on packages/core/src/linker/template_ref.ts:
```diff
@@ -96,6 +99,17 @@ const R3TemplateRef = class TemplateRef<T> extends ViewEngineTemplateRef<T> {
     return new R3_ViewRef<T>(embeddedLView);
   }
+
+  private createEmbeddedViewInjector(embeddedViewInjector: Injector|undefined): Injector|null {
```
Minor optimization proposal: since we don't do property renaming, the code like `this._declarationLView` or method names are not minified, so they'd be retained as is. In order to optimize the code, we can do something like this in the `TemplateRef.createEmbeddedView` function:
```
override createEmbeddedView(context: T, injector?: Injector): EmbeddedViewRef<T> {
  const declarationLView = this._declarationLView;
  // ...
  const embeddedViewInjector = injector ? createChainedInjector(injector, declarationLView[INJECTOR]) : null;
  const embeddedLView = createLView(
        declarationLView, embeddedTView, context, LViewFlags.CheckAlways, null,
        embeddedTView.declTNode, null, null, null, embeddedViewInjector);
  ...
  
```
In which case the `createChainedInjector` might be a generic function that can live outside of a class and can be reused in the future (if needed):
```
function createChainedInjector(ownInjector: Injector, parentInjector?: Injector): Injector {
  return parentInjector ?
      new ChainedInjector(ownInjector, parentInjector) :
      ownInjector;
}
```

crisbeto commented:
The feedback has been addressed.

crisbeto commented:
@AndrewKushnir @jelbourn I've changed the public API based on the discussion. I kept the `injector` positional parameter as well, because it seemed weird to only have the new feature in one signature. Can you take another look?

AndrewKushnir commented on packages/core/src/render3/component_ref.ts:
```diff
@@ -135,11 +114,11 @@ export class ComponentFactory<T> extends viewEngine_ComponentFactory<T> {
 
     const rendererFactory =
-        rootViewInjector.get(RendererFactory2, domRendererFactory3) as RendererFactory3;
+        rootViewInjector.get(RendererFactory2, domRendererFactory3 as RendererFactory2) as
```
Just curious if this extra type cast here (`as RendererFactory2`) is needed as it wasn't there before.

AndrewKushnir replied:
Un-resolving this comment as the type cast has returned :) Do we need it?

crisbeto replied:
We still need the cast, because the injector expects the fallback value to be the same as the value being injected. We didn't need it previously, because it wasn't inferred correctly.

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -3546,4 +3546,391 @@ describe('di', () => {
+      expect(injectedValue).toBe('hello from injector');
+    });
+
+    it('should use a custom injector when the view is created outside of the declaration view',
```
Would it make sense to add a simple test with multi-providers as well?

crisbeto replied:
I'm not sure. The value creation parts are the same as all other DI tests. The main difference here is the injector hierarchy.

AndrewKushnir reviewed (commented):
Looks good (just left 2 minor comments) 👍 

I'm not sure what's the benefit of having 2 APIs to provide an injector: via an options object (3rd argument) and by using positional argument (4th arg). I'd probably vote for having only one option (via options object) if there is no strong reason to add both. We could introduce 4th positional argument later if there is a feedback that the proposed API is not ergonomic.

crisbeto commented:
@AndrewKushnir @jelbourn unfortunately the PR needs to be re-reviewed, because I ended up rewriting the implementation. While adding more unit tests, I noticed that the behavior was a little weird, because the injector of the embedded view's parent component would take precedence over the injector specified when creating the view (see the test called "should have value from custom injector take precedence over parent injector"). The problem is that `LView[INJECTOR]` is used as a module injector which we may never hit if the token exists on a node injector.

AndrewKushnir commented on packages/core/src/render3/di.ts:
```diff
@@ -405,6 +405,14 @@ function lookupTokenUsingModuleInjector<T>(
     flags: InjectFlags = InjectFlags.Default, notFoundValue?: any): T|null {
+  const embeddedInjector = lView[EMBEDDED_INJECTOR];
+  if (embeddedInjector) {
+    const value = embeddedInjector.get(token, NOT_FOUND as T, flags & InjectFlags.Optional);
```
Could you please add a quick comment here to explain why we use `Optional` flag here (so that we don't fail and proceed to the next injector)?

JoostK replied:
Is this really supposed to use `flags & InjectFlags.Optional` instead of `flags | InjectFlags.Optional`? The former reduces `flags` to only retain the `Optional` variant, the latter ensures that `Optional` is always present (and that seems more logical to me here, although a default value is present so I'm not even sure if `Optional` makes any difference)

AndrewKushnir commented on packages/core/src/render3/di.ts:
```diff
@@ -405,6 +405,14 @@ function lookupTokenUsingModuleInjector<T>(
 export function getOrCreateInjectable<T>(
     tNode: TDirectiveHostNode|null, lView: LView, token: ProviderToken<T>,
     flags: InjectFlags = InjectFlags.Default, notFoundValue?: any): T|null {
+  const embeddedInjector = lView[EMBEDDED_INJECTOR];
```
nit: I think it'd be a bit more straightforward if we include `VIEW` into the name (so it's more obvious that this is an injector from an embedded view):
```suggestion
  const embeddedInjector = lView[EMBEDDED_VIEW_INJECTOR];
```

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -3546,4 +3546,430 @@ describe('di', () => {
+         const fixture = TestBed.createComponent(App);
+         fixture.detectChanges();
+
+         fixture.componentInstance.creator.trigger.menu =
+             fixture.componentInstance.declarer.template;
+         fixture.componentInstance.creator.trigger.open(injector);
+         fixture.detectChanges();
```
nit: I think it'd be a bit easier to read if we create a new vars for some values:
```suggestion
         const {creator, declarer} = fixture.componentInstance;
         creator.trigger.menu = declarer.template;
         creator.trigger.open(injector);
         fixture.detectChanges();
```

AndrewKushnir reviewed (approved):
LGTM 👍 

There are some comments, but they are quite minor.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

JoostK commented on packages/core/src/render3/di.ts:
```diff
@@ -405,6 +405,14 @@ function lookupTokenUsingModuleInjector<T>(
     tNode: TDirectiveHostNode|null, lView: LView, token: ProviderToken<T>,
     flags: InjectFlags = InjectFlags.Default, notFoundValue?: any): T|null {
+  const embeddedInjector = lView[EMBEDDED_INJECTOR];
+  if (embeddedInjector) {
```
What would happen for something like this:

```html
<ng-template #a>
  <ng-template #b>
    
  </ng-template>
</ng-template>
```

Should creating a view from `b` have access to an embedded injector of `#a`? I think it should, and by looking at this code I'm not sure if that currently behaves like that: this logic is executed only once in `getOrCreateInjectable` whereas we jump to higher views in the `while` loop below. I think that means that we wouldn't consider the embedded view injector of a parent embedded view.

crisbeto replied:
That's a good observation. I've added some code to handle this case and some more contrived ones.

JoostK replied:
I wonder now how this behaves with respect to providers from the `tNode`, or any `tNode` in between two embedded views.

For example:

```html
<ng-template #a>
  <my-cmp>
    <ng-template #b>
      
    </ng-template>
  </my-cmp>
</ng-template>
```

where `my-cmp` provides some token that is also provided by the injector for view `a`. Currently we eagerly traverse the embedded view injectors _before_ considering any of the nodes, so I'd presume that the token from `a` is found, not the closer one from `my-cmp`.

JoostK replied:
I think we'd need to look into the embedded view container only at the view boundary or something like that to make that work.

crisbeto replied:
Hmm, this is starting to get tricky, because we may end up re-implementing part of the DI system just for the embedded view case. There are other use cases where it's unclear which providers should take precedence, like a directive with `providers` on an `ng-template` that is created using `createEmbeddedView` with an injector. I wonder whether these cases work with `ViewContainerRef.createComponent` which provides the injector in a similar way (except for the part that walks up the parent views) since this is meant to mirror its behavior.

crisbeto replied:
@alxhub, @AndrewKushnir and I had a discussion about this case and concluded to revert to the original behavior where the custom injector acts as a module injector, rather than trying to insert it in the node injector tree. The reasons are:

1. It's consistent with the behavior when an injector is passed in to `ViewContainerRef.createComponent` which `createEmbeddedView` is meant to mirror.
2. It isn't actually clear which node injector tree we would inserting into. If we take the following example:
```
<comp-a>
  <ng-template #template>
    <child-comp></child-comp>
  </ng-template>
</comp-a>

<comp-b>
  <create-embedded-view-here [template]="template"></create-embedded-view-here>
</comp-b>
```

In the example it isn't clear whether the `child-comp` would inherit the providers from the declaration location (`comp-a`) or the insertion location (`comp-b`). A case can be made for both.

jessicajaniuk commented on packages/core/src/render3/interfaces/injector.ts:
```diff
@@ -48,7 +48,7 @@ import {LView, TData} from './view';
  * index + 8: cumulative bloom filter
  * index + TNODE: TNode associated with this `NodeInjector`
- *                `canst tNode = tView.data[index + NodeInjectorOffset.TNODE]`
+ *                `const tNode = tView.data[index + NodeInjectorOffset.TNODE]`
```
I canst believe this spelling error was here. Good catch!

jessicajaniuk reviewed (approved):
LGTM :cookie:

Reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api

crisbeto commented:
Reverted to the original based on the discussion @AndrewKushnir.

AndrewKushnir commented on packages/core/src/render3/chained_injector.ts:
```diff
@@ -0,0 +1,36 @@
+ * injector. Used primarily when creating components or embedded views dynamically.
+ */
+export class ChainedInjector implements Injector {
+  constructor(private injector: Injector, private moduleInjector: Injector) {}
```
Can we rename `moduleInjector` -> `parentInjector`? From the `ChainedInjector` class perspective, there is no notion of module injector, it's just some other injector :)
```suggestion
  constructor(private injector: Injector, private parentInjector: Injector) {}
```

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -3546,4 +3546,599 @@ describe('di', () => {
+         const fixture = TestBed.createComponent(App);
+         fixture.detectChanges();
+
+         fixture.componentInstance.outerTrigger.open(
+             Injector.create({providers: [{provide: token, useValue: 'hello'}]}));
+         fixture.detectChanges();
+
+         fixture.componentInstance.innerTrigger.open(undefined);
+         fixture.detectChanges();
+
+         expect(fixture.componentInstance.innerMenu.tokenValue).toBe('hello');
```
nit: can we create a separate consts for the `fixture.componentInstance` fields, so the code is a bit more readable:
```suggestion
         const {outerTrigger, innerTrigger, innerMenu} = fixture.componentInstance;
         outerTrigger.open(
             Injector.create({providers: [{provide: token, useValue: 'hello'}]}));
         fixture.detectChanges();

         innerTrigger.open(undefined);
         fixture.detectChanges();

         expect(innerMenu.tokenValue).toBe('hello');
```

crisbeto replied:
The problem is that `innerTrigger` and `innerMenu` don't exist until `outerTrigger.open` is called.

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -3546,4 +3546,599 @@ describe('di', () => {
+             fixture.componentInstance.openFromComponent(Injector.create({providers}));
+         fixture.detectChanges();
+
+         expect(fixture.componentInstance.overlayInTemplate.tokenValue).toBe('hello from parent');
+         expect(componentRef.instance.overlay.tokenValue).toBe('hello from parent');
```
Could you please add a comment to explain why it's expected that our provided Injector would not take precedence?

AndrewKushnir reviewed (approved):
LGTM 👍 Just left a few comments, but they are pretty minor.

crisbeto committed:
feat(core): allow for injector to be specified when creating an embedded view

Adds support for passing in an optional injector when creating an embedded view through `ViewContainerRef.createEmbeddedView` and `TemplateRef.createEmbeddedView`. The injector allows for the DI behavior to be customized within the specific template.

Fixes #14935.

dylhunn commented:
This PR was merged into the repository by commit b49ffcd50e46ca7907ee2049f6cade8e080f5980.
